F/copyright review 2015 extended

bkil-syslogng opened the pull request:
This is the extended version of the copyright cleanup PR #822. Theoretically all previous occurrences have been replaced with the new convention of Copyright (c) 2015 Balabit, but do comment if you this it has an error. It is enough to merge f/copyright-review-2015-extended, because it also contains f/copyright-review-2015.

Fixes #824

bazsi commented:
This looks good to me.

I was thinking that I saw too many incorrect license statements in the tree already, so we should add some kind of automated copyright checks and a written policy which license to apply where.

bkil-syslogng commented:
I've rebased PR #823 f/copyright-review-2015-extended to master (now that the date fixes have been merged). Travis should now pass, and single click merge should work. I'll probably close #822 non-extended then if this version is okay with you.

bkil-syslogng commented:
@bazsi In the last separate commit, we have relicensed modules/java and modules/java-modules/common under the LGPL. Please verify whether this is how you want it to be. Nobody has reviewed my changes yet, but I have consulted with Laci, Viktor and Tibi beforehand. I could rebase some of the changes to the modules/java-modules/common commit if you wish, although the relicensing itself is considered a separate and not insignificant action in my point of view, so I'd leave it as it is.

bazsi commented:
It should be separate. To make reviewing easier and automation possible at a future point, could you pls summarize the current licensing in a gitignore style file?

Eg. each line two whitespace separated columns, first the wildcard pattern, second the intended license (first match wins). This should clearly specify what you intended and would make review easier.

Apart from that I would definitely go for merging the extended variant, perhaps split the PR and merge it split by split as it can deteriorate pretty quickly.

Thanka

bazsi commented:
I have now reviewed the entire series, and it is good to go.  Thanks a lot.

The license listing would still make sense to do until we remember it. While I was at it, I quickly collected the stuff I know:

```
bazsi@bzorp:~/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng$ cat license-layout.txt 
lib/*				LGPL
libtest/*			LGPL
syslog-ng/*			LGPL
modules/java-modules/common/*	LGPL
modules/java/native/*		LGPL
modules/java/proxies/*		LGPL
modules/java/src/*		LGPL
modules/*			GPL
scl/*                                 GPL
```

bkil-syslogng commented:
@bazsi Thanks for the effort. Actually I wanted to do that today, but I see you were faster. Anyway, it's not a bad thing that you have a grand overview of this important aspect. Now that we have declared this policy, if all agree to this scheme, some more license declarations will need fixing. I found occurrences of an invalid declaration where LGPL/GPL was referenced at the same time, and now I have the information to decide which one to reword.

lbudai commented:
@bkil-syslogng : will you update this PR with other license fixes? or you will create another PR for those?

bkil-syslogng commented:
@lbudai I'd like to update this one. Although, hundreds of files have been touched by the above fixes already in a mostly orthogonal manner, a hundred more changes are in order, mostly of similar intention and kind (copyright declarations).

bkil-syslogng commented:
@bazsi:
Automated checking is almost ready, and I'm still working on that, and will eventually plug it into Travis as well. More than 100 files are still in error, and I'm just starting to plan next steps accordingly. A few mistakes are obvious, while other occasions may have been intentional.

From a legal perspective, I would need to do some detective work from git and do discussions whether we should consider these relicensing, or just correction of obvious errors.

I basically found the following licenses: GPLv2+_SSL, GPLv2+, LGPLv2.1+, LGPLv2.1+_SSL and the newly found code parts. Are all supposed to be the SSL variant? Or should we give thought about which parts will actually link to SSL and only license thoso? Is it okay if certain modules and files are LGPL, while they should have been GPL according to the tree described above? Anyway, I think they are convertible in the following direction:
LGPLv2.1+_SSL -> GPLv2+_SSL, GPLv2+, LGPLv2.1+
LGPLv2.1+ -> GPLv2+
GPLv2+_SSL -> GPLv2+

Just for the record, I'm using the following rules file at the moment (my own format):

 LGPLv2.1+_SSL
lib/(compat|control|debugger|filter|logproto|parser|rewrite|stats|template|tests|transport|[^/]*$)
libtest
syslog-ng(-ctl)?
modules/java-modules/common
modules/java/(native|proxies|src)
 GPLv2+_SSL
tests
modules/java/(tools|[^/]*$)
modules/java-modules/(dummy|elastic|hdfs|http|kafka|[^/]*$)
modules/(afamqp|affile|afmongodb|afprog|afsmtp|afsocket|afsql|afstomp|afstreams|afuser|basicfuncs|confgen|cryptofuncs|csvparser|date|dbparser|graphite|json|kvformat|linux-kmsg-format|pacctformat|pseudofile|python|redis|riemann|syslogformat|systemd-journal|system-source|[^/]*$)
modules/geoip/(tfgeoip\.c|Makefile\.am)
scl
 GPLv2+_SSL,non-balabit
modules/geoip/geoip-(plugin\.c|parser(|-parser)\.[ch])

By the way, do you happen to know whether modules/geoip was created in Balabit worktime?

bkil-syslogng commented:
Investigating `modules/geoip/geoip-(plugin\.c|parser(|-parser)\.[ch])` contributed by algernon in his free time made me think about this in general.

On second thought, regardless of the value added by Balabit, all syslog-ng code should carry an additional (c) Balabit line as well as their original authors to indicate our blessing of the functionality and having and keeping it integrated. You could think of this as an 'instant and automatic fork' of every codebase we receive. This should be allowed by the respective licenses. I'm adding the extra lines in these few files if there are no objections.

bazsi commented:
Hi,

On Fri, Dec 11, 2015 at 10:41 AM, Tamas Nagy <notifications@github.com>
wrote:

> @bazsi <https://github.com/bazsi>:
> Automated checking is almost ready, and I'm still working on that, and
> will eventually plug it into Travis as well. More than 100 files are still
> in error, and I'm just starting to plan next steps accordingly. A few
> mistakes are obvious, while other occasions may have been intentional.
>
Thanks a lot for this effort. I think the licensing stuff of syslog-ng is
complex enough, so automated testing against an easier to verify source of
truth would help a lot.


> From a legal perspective, I would need to do some detective work from git
> and do discussions whether we should consider these relicensing, or just
> correction of obvious errors.
>

If a code was contributed by Balabit then the question is really simple. If
however a file has copyrights from multiple parties and the change we want
to do is a real logical change, we need to contact the author(s).


> I basically found the following licenses: GPLv2+_SSL, GPLv2+, LGPLv2.1+,
> LGPLv2.1+_SSL and the newly found code parts. Are all supposed to be the
> SSL variant?
>
yes, for safety. I would add the SSL exemption for code where we can decide
on our own or all the authors are readily available. Elsewhere if the code
is not clearly dependant on OpenSSL and we don't have a contact with the
original author I would leave it as is.

Let's see how much stuff remains and decide then how much time it would
need to fix the situation.


> Or should we give thought about which parts will actually link to SSL and
> only license thoso? Is it okay if certain modules and files are LGPL, while
> they should have been GPL according to the tree described above?
>
It would be nice if you could point out these issues, but LGPL can always
be converted to GPL, that's explicitly allowed by the license.


> Anyway, I think they are convertible in the following direction:
> LGPLv2.1+_SSL -> GPLv2+_SSL, GPLv2+, LGPLv2.1+
> LGPLv2.1+ -> GPLv2+
> GPLv2+_SSL -> GPLv2+
>
I would try to simplify the licenses as much as possible and use SSL
everywhere, except if it's not there and is too much effort to put it there.

> Just for the record, I'm using the following rules file at the moment (my
> own format):
>
> LGPLv2.1+_SSL
>
> lib/(compat|control|debugger|filter|logproto|parser|rewrite|stats|template|tests|transport|[^/]
>
>
>
>
>
>
> *$) libtest syslog-ng(-ctl)? modules/java-modules/common
> modules/java/(native|proxies|src) GPLv2+_SSL tests modules/java/(tools|[^/]*
> $)
> modules/java-modules/(dummy|elastic|hdfs|http|kafka|[^/]
> *$)
> modules/(afamqp|affile|afmongodb|afprog|afsmtp|afsocket|afsql|afstomp|afstreams|afuser|basicfuncs|confgen|cryptofuncs|csvparser|date|dbparser|graphite|json|kvformat|linux-kmsg-format|pacctformat|pseudofile|python|redis|riemann|syslogformat|systemd-journal|system-source|[^/]*
> $)
> modules/geoip/(tfgeoip.c|Makefile.am)
> scl
> GPLv2+_SSL,non-balabit
> modules/geoip/geoip-(plugin.c|parser(|-parser).[ch])
>
I wouldn't mix the copyright attribution with the license. They should be
separated, and I think license validation is the more important. copyright
attribution is usually ok, because we tend to add those at the initial
merge and they don't change. licensing however is easy to mix up, LGPL
looks awfully similar to GPL. :)

> By the way, do you happen to know whether modules/geoip was created in
> Balabit worktime?
>

As I see there both outside patches (Algernon signed off these patches as @
madhouse.org and not from balabit) and BB ones.
Bazsi

bazsi commented:
See inline.

On Fri, Dec 11, 2015 at 2:27 PM, Tamas Nagy <notifications@github.com>
wrote:

> Investigating modules/geoip/geoip-(plugin\.c|parser(|-parser)\.[ch])
> contributed by algernon in his free time made me think about this in
> general.
>
> On second thought, regardless of the value added by Balabit, all syslog-ng
> code should carry an additional (c) Balabit line as well as their original
> authors to indicate our blessing of the functionality and having and
> keeping it integrated. You could think of this as an 'instant and automatic
> fork' of every codebase we receive. This should be allowed by the
> respective licenses. I'm adding the extra lines in these few files if there
> are no objections.
>
I wouldn't do that, and only add the BB copyright in case there's a
copyrightable contribution by BB. Merging it in a codebase is not a
contribution, neither is it an "official blessing".

Of course if we helped the implementation with a sizable review feedback
and/or provided changes/fixes during the implementation, then we could
easily add that copyright statement. It is really not nice to add this
automatically at merging though, if it was merged without changes. I
wouldn't be happy about this if someone did this to my patches in other
projects.



> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/823#issuecomment-163935914>.
>

bkil-syslogng commented:
What is the future of the AUTHORS file? debian/copyright doesn't look too good either, however the Ubuntu package I've checked contain some pretty nice ones, maybe we should consider updating based on that. We should probably greatly extend the COPYING file based on that and how others solve the question. I've just checked the ChromeCast one, and they use a very long, but semi-automatically generated one which looks professional.

We should probably decide whether we support blanket AUTHORS-mention kind of contributions as well. I don't recommend that, and actually I would move the respective copyrights from AUTHORS to the respective files instead.

Although AUTHORS can reduce per-file copyright tracking overhead, it could really make your life miserable in case of relicensing and legal troubles.

I've proposed a licensing (copyright?) section for the `CONTRIBUTORS` file. What do you think?

bkil-syslogng commented:
I would consider unification of the identities of the copyright holders. This usually means that if you've changed e-mail addresses in the timeframe of your contributions eligible for copyright, all of the given addresses should be listed at each occurrences by default. This can matter in case of common names, publications and derived works. I guess this usually boils down to personal choice, so a given author could choose to stick to specifying the following e-mail address:
* none (default if none given);
* the first one;
* the most recent one;
* all ones used in the code in the contribution time frame (default).

bkil-syslogng commented:
>
> See inline.
>
> On Fri, Dec 11, 2015 at 2:27 PM, Tamas Nagy <notifications@github.com>
> wrote:
>
> > Investigating modules/geoip/geoip-(plugin\.c|parser(|-parser)\.[ch])
> > contributed by algernon in his free time made me think about this in
> > general.
> >
> > On second thought, regardless of the value added by Balabit, all
> syslog-ng
> > code should carry an additional (c) Balabit line as well as their
> original
> > authors to indicate our blessing of the functionality and having and
> > keeping it integrated. You could think of this as an 'instant and
> automatic
> > fork' of every codebase we receive. This should be allowed by the
> > respective licenses. I'm adding the extra lines in these few files if
> there
> > are no objections.
> >
> I wouldn't do that, and only add the BB copyright in case there's a
> copyrightable contribution by BB. Merging it in a codebase is not a
> contribution, neither is it an "official blessing".
>
> Of course if we helped the implementation with a sizable review feedback
> and/or provided changes/fixes during the implementation, then we could
> easily add that copyright statement. It is really not nice to add this
> automatically at merging though, if it was merged without changes. I
> wouldn't be happy about this if someone did this to my patches in other
> projects.


Alright, we could stick to not adding an extra Balabit line, and reverting
the said addition. This sounded morally the best for me at first, however
we will eventually touch every single file very soon, I viewed this more of
a technical issue, than a moral or legal one. I was only concerned about
whether this would bring in too much of an overhead.

As you can see in the above cited configuration file, I already have
automatic checking in place for misattribution as well. If your above
question was regarding this (the non-balabit tag), then I would keep this
checking enabled for the time being. The situation when a file has not been
touched creatively by Balabit is very rare, in exchange for this single
line of annotation (and a few more folders under this heading as new
modules will be added), we've found lots of cases of misattribution and
confusion. I call that a great bang for the buck. If your question was
about my focus, misattribution in general has taken only a tiny fraction of
my time until now (and I think their respective count is now at 0).

Do note that the git tree won't contain information about the review
process, so in the case that you describe, I'd agree with the PR owner that
he/she should add our copyright at the end if we all feel that we have
produced value together. We should probably add this to the CONTRIBUTING
guide.

I do have a good reason to keep the files as marked as cleanly as possible.
On VCS import/exports, manual cherry picking, restructuring and some other
occasions, you sometimes (almost always) lost attributions, which slows me
down tracking down the list of contributors. And this is in our present
fortunate case where we still have access to most of the trees. I've even
found cases where the last copyright review from many years ago introduced
misattributions because they were not following along lineage. Although
this is more of a nitpick than an issue, it points out how difficult this
is to get right afterwards.

bazsi commented:
Hmmm... the AUTHORS file is more a kind of  "credits" file, it doesn't need
frequent updates but I'd keep that.

The debian directory is more like a placeholder, there is PR #378 that
should be discussed, concluded and that should solve the debian directory
issues.

If we have a file that is automatically checked, that should be mentioned
in COPYING instead of trying to translate that to natural language.

I am reviewing the CONTRIBUTORS changes as review comments.

-- 
Bazsi

On Sat, Dec 12, 2015 at 3:11 PM, Tamas Nagy <notifications@github.com>
wrote:

> What is the future of the AUTHORS file? debian/copyright doesn't look too
> good either, however the Ubuntu package I've checked contain some pretty
> nice ones, maybe we should consider updating based on that. We should
> probably greatly extend the COPYING file based on that and how others solve
> the question. I've just checked the ChromeCast one, and they use a very
> long, but semi-automatically generated one which looks professional.
>
> We should probably decide whether we support blanket AUTHORS-mention kind
> of contributions as well. I don't recommend that, and actually I would move
> the respective copyrights from AUTHORS to the respective files instead.
>
> Although AUTHORS can reduce per-file copyright tracking overhead, it could
> really make your life miserable in case of relicensing and legal troubles.
>
> I've proposed a licensing (copyright?) section for the CONTRIBUTORS file.
> What do you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/823#issuecomment-164157489>.
>

bkil-syslogng commented:
>
> Hmmm... the AUTHORS file is more a kind of "credits" file, it doesn't need
> frequent updates but I'd keep that.
>
> This may seem like a stupid question, but what kind of file is a credits
file then? If you would like to give full credit, as a first measure, you
could simply concatenate all copyright holders, which could also be
automated.

Of course, giving credit is subjective, because sometimes copy & pasting
whole files seems trivial. At other times a few lines of modifications may
have fixed some very deeply lurking issue in the architecture which has
contributed great insight. Although, we shouldn't forget the case when a
given valuable contribution was mostly removing items (lines, or files),
which in the end may remove all artifacts to which we attach declarations.

So on second thought, we do need some kind of extra AUTHORS/credits file
for the above case in which all traces of contributions have been deleted.
I would probably list all past contributors and those deduced from
declarations present in live code.

The debian directory is more like a placeholder, there is PR #378 that
> should be discussed, concluded and that should solve the debian directory
> issues.
>
> That would be great. Otherwise it gives a nice overview of our licensing
policy.

If we have a file that is automatically checked, that should be mentioned
> in COPYING instead of trying to translate that to natural language.


> Maybe I don't have a good enough understanding of this sentence. Do you
suggest that COPYING should link to this LICENSE.policy file (naming: TODO)
and we shall distribute both together? I only copied the relevant parts to
the COPYING file because that is what we distribute and because it is a
conventional choice. Or perhaps would you like to automatically generate a
COPYING file by concatenating a COPYING.template and LICENSE.policy?


> I am reviewing the CONTRIBUTORS changes as review comments.
>
> Thanks.

bkil-syslogng commented:
@bazsi I'm ready with the review and fixing part. I'm moving onto automation. Here are my findings that we should address ASAP:

problematic incompatible deviation:
error: mismatch lib/filter/filter-netmask6.c expected:LGPLv2.1+_SSL detected:GPLv2+_SSL
error: mismatch lib/filter/filter-netmask6.h expected:LGPLv2.1+_SSL detected:GPLv2+_SSL
error: mismatch lib/host-id.c expected:LGPLv2.1+_SSL detected:GPLv2+_SSL
error: mismatch lib/host-id.h expected:LGPLv2.1+_SSL detected:GPLv2+_SSL
error: mismatch lib/stats/stats-counter.c expected:LGPLv2.1+_SSL detected:GPLv2+_SSL
error: mismatch lib/template/common-template-typedefs.h expected:LGPLv2.1+_SSL detected:GPLv2+_SSL

nuisance incompatible deviation:
error: mismatch lib/logproto/logproto-indented-multiline-server.c expected:LGPLv2.1+_SSL detected:LGPLv2.1+
error: mismatch lib/logproto/logproto-indented-multiline-server.h expected:LGPLv2.1+_SSL detected:LGPLv2.1+
error: mismatch lib/logproto/logproto-regexp-multiline-server.c expected:LGPLv2.1+_SSL detected:LGPLv2.1+
error: mismatch lib/logproto/logproto-regexp-multiline-server.h expected:LGPLv2.1+_SSL detected:LGPLv2.1+
error: mismatch modules/linux-kmsg-format/linux-kmsg-format.c expected:GPLv2+_SSL detected:GPLv2+
error: mismatch modules/linux-kmsg-format/linux-kmsg-format.h expected:GPLv2+_SSL detected:GPLv2+
error: mismatch modules/linux-kmsg-format/linux-kmsg-format-plugin.c expected:GPLv2+_SSL detected:GPLv2+
error: mismatch modules/linux-kmsg-format/tests/test_linux_format_kmsg.c expected:GPLv2+_SSL detected:GPLv2+

compatible deviation:
error: mismatch modules/basicfuncs/tf-template.c expected:GPLv2+_SSL detected:LGPLv2.1+_SSL
error: mismatch modules/python/python-config.c expected:GPLv2+_SSL detected:LGPLv2.1+_SSL
error: mismatch modules/python/python-debugger.c expected:GPLv2+_SSL detected:LGPLv2.1+_SSL
error: mismatch modules/python/python-debugger.h expected:GPLv2+_SSL detected:LGPLv2.1+_SSL
error: mismatch modules/python/python-helpers.c expected:GPLv2+_SSL detected:LGPLv2.1+_SSL
error: mismatch modules/python/python-helpers.h expected:GPLv2+_SSL detected:LGPLv2.1+_SSL
error: mismatch modules/python/python-logmsg.c expected:GPLv2+_SSL detected:LGPLv2.1+_SSL
error: mismatch modules/python/python-logmsg.h expected:GPLv2+_SSL detected:LGPLv2.1+_SSL
error: mismatch modules/python/python-main.c expected:GPLv2+_SSL detected:LGPLv2.1+_SSL
error: mismatch modules/python/python-module.h expected:GPLv2+_SSL detected:LGPLv2.1+_SSL
error: mismatch modules/python/python-tf.c expected:GPLv2+_SSL detected:LGPLv2.1+_SSL
error: mismatch modules/python/python-value-pairs.c expected:GPLv2+_SSL detected:LGPLv2.1+_SSL

bkil-syslogng commented:
I have finished development of the checker. I have squashed commits and rebased to master. Someone please do a quick sanity check on the end result and merge ASAP.

bkil-syslogng commented:
I've rebased to master (after CEF merge)

bkil-syslogng commented:
I've rebased to master (following misc refactor, suspend if queue, make func tests more flexible)

bkil-syslogng commented:
We've agreed with @ihrwein that he'll check out and review this final PR in depth now and if he likes it, someone should merge it.

ihrwein commented:
:+1: (I'm not a sed expert and the last commit is a little bit sed heavy, so that should be checked by the other devs)

bkil-syslogng commented:
We have discussed with @ihrwein  and @bazsi  to relicense `modules/native` under the LGPL:
https://github.com/bkil-syslogng/syslog-ng/commit/b7e5b75c5b76c88979c503c2ba8fa4e5ac1ada99

I have then rebased this again to master (after the `-Werror` merge). I have backed up the branch before rebasing: https://github.com/bkil-syslogng/syslog-ng/pull/18

Travis still passes.

Still waiting for an `all clear` signal from `somebody`.

ihrwein commented:
maybe @bazsi can press the button :)

bazsi commented:
thanks for the comprehensive review and the solution to automate these checks.
